packages/gatsby/src/internal-plugins/query-runner/__tests__/pages-writer.js

```diff
@@ -0,0 +1,88 @@
+const _ = require(`lodash`)
+const { writePages, resetLastHash } = require(`../pages-writer`)
+
+const jsonDataPathsFixture = require(`./fixtures/jsonDataPaths.json`)
+const pagesFixture = require(`./fixtures/pages.json`)
+
+jest.mock(`fs-extra`, () => {
+  return {
+    writeFile: (filePath, data) => Promise.resolve(),
+    move: () => {},
+  }
+})
+const mockFsExtra = require(`fs-extra`)
+
+let mockState = {
+  program: {
+    _: [`build`],
+    directory: `my/gatsby/project`,
+  },
+  jsonDataPaths: jsonDataPathsFixture,
+  pages: {
+    values: () => pagesFixture,
+  },
+}
+
+jest.mock(`../../../redux/`, () => {
+  return {
+    store: {
+      getState: () => mockState,
+    },
+    emitter: {
+      on: () => {},
+    },
+  }
+})
+
+describe(`Pages writer`, () => {
+  it(`writes pages with the good order`, async () => {
+    const spy = jest.spyOn(mockFsExtra, `writeFile`)
+
+    await writePages()
+    const data1 = spy.mock.calls[3][1]
+
+    mockState = {
+      ...mockState,
+      jsonDataPaths: {
+        bar: jsonDataPathsFixture.bar,
+        foo: jsonDataPathsFixture.foo,
+        baz: jsonDataPathsFixture.baz,
+      },
+      pages: {
+        values: () =>
+          _(pagesFixture)
+            .chunk(2)
+            .reverse()
+            .flatten()
+            .value(),
+      },
+    }
+
+    // Ensure testing in the same conditions as if we have
+    // removed the .cache folder
+    resetLastHash()
+
+    await writePages()
+    const data2 = spy.mock.calls[7][1]
+
+    expect(spy.mock.calls.length).toBe(8)
```

Comment:
Where's this "magic number" coming from? Can we infer `.length` from something?

Also quick tip :)

```js
expect(spy).toHaveBeenCalledTimes(8);
```

🎉

Reply:
It comes from this [4 lines](https://github.com/gatsbyjs/gatsby/blob/fcef8b3e6ea92547185c26ce378a65e007611106/packages/gatsby/src/internal-plugins/query-runner/pages-writer.js#L110-L113)

Thanks for the tip! 😉